assets/src/dashboard/components/infiniteScroller/index.js

```diff
@@ -21,82 +21,128 @@ import { __ } from '@wordpress/i18n';
 /**
  * External dependencies
  */
-import {
-  useState,
-  useEffect,
-  useLayoutEffect,
-  useCallback,
-  useRef,
-} from 'react';
+import { useEffect, useRef, useReducer } from 'react';
 import PropTypes from 'prop-types';
+import styled from 'styled-components';
+
+// TODO
+const ScrollMessage = styled.div`
+  margin: 20px 0;
+  text-align: center;
+  width: 100%;
+`;
+
+const STATE = {
+  loadable: 'loadable',
+  loading_internal: 'loading_internal',
+  loading_external: 'loading_external',
+  complete: 'complete',
+};
+
+const ACTION = {
+  ON_INTERNAL_LOAD: 'internal on load',
+  ON_EXTERNAL_LOAD: 'external on load',
+  ON_ALL_LOADED: 'all loaded',
+  ON_LOAD_SUCCESS: 'load success',
+};
+
+const machine = {
+  [STATE.loadable]: {
+    [ACTION.ON_INTERNAL_LOAD]: STATE.loading_internal,
+    [ACTION.ON_EXTERNAL_LOAD]: STATE.loading_external,
+  },
+  [STATE.loading_internal]: {
+    [ACTION.ON_ALL_LOADED]: STATE.complete,
+    [ACTION.ON_LOAD_SUCCESS]: STATE.loadable,
+  },
+  [STATE.loading_external]: {
+    [ACTION.ON_ALL_LOADED]: STATE.complete,
+    [ACTION.ON_LOAD_SUCCESS]: STATE.loadable,
+  },
+  [STATE.complete]: {
+    [ACTION.ON_LOAD_SUCCESS]: STATE.loadable,
+  },
+};
+
+const loadReducer = (state, action) => {
+  return machine[state][action] || state;
+};
 
 const InfiniteScroller = ({
   allDataLoadedMessage = __('No More Stories', 'web-stories'),
-  children,
-  handleGetData,
-  isAllDataLoaded,
+  onLoadMore,
+  canLoadMore,
+  isLoading,
   loadingMessage = __('Loading…', 'web-stories'),
 }) => {
-  const [loadMore, setLoadMore] = useState(false);
   const loadingRef = useRef(null);
-  const containerRef = useRef(null);
+  const onLoadMoreRef = useRef(onLoadMore);
+  onLoadMoreRef.current = onLoadMore;
+
+  const [loadState, dispatch] = useReducer(loadReducer, STATE.loadable);
+
+  useEffect(() => {
+    if (loadState === STATE.loading_internal) {
+      onLoadMoreRef.current();
+    }
+  }, [loadState]);
+
+  useEffect(() => {
+    if (isLoading) {
+      dispatch(ACTION.ON_EXTERNAL_LOAD);
+    }
+  }, [isLoading]);
+
+  useEffect(() => {
+    if (!isLoading) {
+      dispatch(canLoadMore ? ACTION.ON_LOAD_SUCCESS : ACTION.ON_ALL_LOADED);
+    }
+  }, [isLoading, canLoadMore]);
+
   useEffect(() => {
-    if (loadMore) {
-      handleGetData(loadMore);
-      setLoadMore(false);
+    if (loadState === STATE.complete && canLoadMore) {
+      dispatch(ACTION.ON_LOAD_SUCCESS);
```

Comment:
why do we want to do anything once we're in the `complete` state?

Also as a side note, you shouldn't need to wrap the dispatch in a state check cus that's what the state machine is already built to do. If you're having clashes with actions that are accepted by multiple states, consider naming it a different action